### PR TITLE
Sanitation Helper

### PR DESCRIPTION
**Pull request by c-butcher**
Added a SanitationHelper along with Sanitizers for booleans, integers, floats, and strings.

### CODE REVIEW

**Comment by AndyM84** [comment]
Though I like the thought of having an interface that defines the basic signature of a sanitizer, I don't like that we have a class which tries to bring them together.  I don't think we can have a hard-coded SanitizerHelper that also allows registration of new types.  It makes them unequal citizens in the land of sanitization, and it just feels off to me.  Will share more thoughts when I'm home.

**Comment by c-butcher** [comment]
There are a few places that I can think of where there will be a lot of input sanitation, and that would be the FormHelper, RequestHelper, CliHelper, and possibly the DatabaseHelper. The major benefit of the SanitationHelper is that it stores and shares a single instance of each sanitizer, which helps to reduce the number of sanitizer instances that are loaded into memory.

I total agree that right now that the Sanitizers are not equal citizens. We could always remove the string(), float(), integer() and boolean() methods which promote our built-in sanitizers as upper-class citizens.

**Comment by c-butcher** [comment]
Alright, I've thought about it more, and it does seem useless to have the SanitationHelper. If we want to save memory, we could create a SanitizerBase and make it a singleton.

**Comment by c-butcher** [comment]
We could also make the Sanitizers have static methods:

IntegerSanitizer::sanitize('123');
BooleanSanitizer::sanitize('false');

File: SanitationHelper.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		 *
+		 * @return mixed
+		 */
+		public function sanitize($input, $class) {
+			if (!isset(self::$_sanitizers[$class])) {
+				if (!class_exists($class)) {
+					return $input;
+				}
+
+				if (!class_implements($class, SanitizerInterface::class)) {
```
Comment (AndyM84) [inline SanitationHelper.php]: This and line 104 (`if (!class_exists($class))`) should be called when a new sanitizer is added, they shouldn't be called each time a request is made.

**Comment by AndyM84** [comment]
I think I'm leaning toward your last comment.  Ideal world means that we're using generics, but we don't have that option here, so maybe if I think of the `CliHelper` as an example, I might expect to see something like this:

```php
use Stoic\Input\Sanitize;

// Create the helper and add the sanitizer(s) I want available
$ch = new CliHelper($argc, $argv);
$ch->addSanitizer('bool', new BooleanSanitizer());

// Check if an argument is available by short or long name
if ($ch->hasParam('d', 'debug')) {
  // Retrieve it using the 'bool' key so we know we're getting it through the sanitizer we supplied
  $debug = $ch->getSanitized('bool', 'd', 'debug');
}
```

It's not perfect, but again it's hard without generics to do something I'd call elegant.

**Comment by c-butcher** [comment]
Alright, we talked about it and decided the SanitationHelper should only be responsible for managing and executing sanitation, but shouldn't come with pre-built sanitation methods. Changes have been added to that effect.

File: SanitationHelper.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+		/**
+		 * Check whether a sanitation type exists.
+		 *
+		 * @param string $key The key, which is a string value, of the sanitizer that is being searched for.
+		 *
+		 * @return bool
+		 */
+		public function hasSanitizer($key) {
+			return isset($this->_sanitizers[$key]);
```
Comment (AndyM84) [inline SanitationHelper.php]: Do we care if we have something set with a `null` value?  It shouldn't be possible with the base class, but what if someone extends and does their own magic with `$this->_sanitizers[]`?

**Comment by c-butcher** [comment]
If the sanitizer is a null value, then obviously no sanitizer exists. If another developer wants to do their own magic and allow for null values in the sanitizers, then they'll need to override the hasSanitizer() method also.

**Pull request merged**